### PR TITLE
[ENH] Implementation of a "complete" FCI algorithm (augmented FCI)

### PR DESCRIPTION
**Pull request by adam2392**
Closes #32 

~Should be merged and rebased after #42 is merged.~

Changes proposed in this pull request:

- implements the FCI algorithm on top of the skeleton learning with latent confounders algorithm implemented in #50 
- This is a complete algorithm that implements R8-10, which is not present in causal-learn
- By changing the line within `LearnSkeleton` to `continue` instead of `break`, this fixes the issue with learning skeletons with PDS being "flaky" https://github.com/py-why/dodiscover/blob/c60cf2ef3a6b5dd05b0d5ad792ce27f1c150f695/dodiscover/constraint/skeleton.py#L388

Note: this is a relatively "large PR" simply because there is a lot of code that goes into implementing each of the rules of the FCI algorithm compared to e.g. the PC algorithm. Moreover, there is 500 LOC for unit tests. Lmk if you spot areas to trim.

TODO:
- [ ] It would be nice to demonstrate to the team in an integration test that R8-10 is able to orient more edges than causal-learn.

## Before submitting

- [ ] I've read and followed all steps in the [Making a pull request](https://github.com/py-why/dodiscover/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [ ] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/py-why/dodiscover/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [ ] If this PR fixes a bug, I've added a test that will fail without my fix.
- [ ] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.

## After submitting

- [ ] All GitHub Actions jobs for my pull request have passed.

### CODE REVIEW

File: dodiscover/constraint/fcialg.py [hunk 1]
```diff
@@ -0,0 +1,725 @@
...
+        # Update the Context:
+        # add the corresponding intermediate PAG now to the context
+        # new initialization graph
+        context.add_state_variable("PAG", pag)
+        for (_, _, d) in new_init_graph.edges(data=True):
+            if "test_stat" in d:
+                d.pop("test_stat")
+            if "pvalue" in d:
+                d.pop("pvalue")
+        context.init_graph = new_init_graph
```
Comment (adam2392) [inline dodiscover/constraint/fcialg.py:686]: See here.

File: dodiscover/context.py [hunk 2]
```diff
@@ -9,6 +9,9 @@
 class Context:
     """Context of assumptions, domain knowledge and data.
 
+    This should NOT be instantiated directly. One should instead
```
Comment (darthtrevino) [inline dodiscover/context.py:12]: To that end, should we continue exporting context in dodiscover/__init__.py, or should we only expose the `make_context` method?

**Review by darthtrevino — DISMISSED** [review]

File: dodiscover/context.py (hunk 2 above)
Comment (adam2392) [reply dodiscover/context.py:12]: Hmm good point. I'll remove from top level import and document that.

**Review by darthtrevino — DISMISSED** [review]
:shipit:

File: tests/unit_tests/constraint/test_fcialg.py [hunk 3]
```diff
@@ -0,0 +1,560 @@
...
+
+        self.G = G
+        self.ci_estimator = oracle
+        self.alg = fci
+
+    def test_fci_skel_graph(self):
+        sample = dummy_sample(self.G)
+        context = make_context().variables(data=sample).build()
+        skel_graph, _ = self.alg.learn_skeleton(sample, context)
+        assert nx.is_isomorphic(skel_graph, nx.Graph([("x", "y"), ("z", "y")]))
```
Comment (robertness) [inline tests/unit_tests/constraint/test_fcialg.py:31]:
```suggestion
        assert nx.is_isomorphic(skel_graph, self.G.to_undirected()))
```

**Review by robertness — APPROVED** [review]
LGTM.  I'll look into working up a notebook

**Pull request merged**